go/vt/mysqlctl/schema.go

```diff
@@ -533,6 +532,79 @@ func (mysqld *Mysqld) ApplySchemaChange(ctx context.Context, dbName string, chan
+// no defined PRIMARY KEY. It will return the columns in a
+// viable PRIMARY KEY equivalent (PKE) -- a NON-NULL UNIQUE
+// KEY -- in the specified table. When multiple PKE indexes
+// are available it will attempt to choose the most efficient
+// one based on the column data types and the number of columns
+// in the index. See here for the data type on disk sizes:
+//   https://dev.mysql.com/doc/refman/en/storage-requirements.html
+// If this function is used on a table that DOES have a
+// defined PRIMARY KEY then it may return the columns for
+// that index if it is likely the most efficient one amongst
+// the  available PKE indexes on the table.
+func (mysqld *Mysqld) GetPrimaryKeyEquivalentColumns(ctx context.Context, dbName, table string) ([]string, error) {
+	conn, err := getPoolReconnect(ctx, mysqld.dbaPool)
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Recycle()
+
+	// We use column name aliases to guarantee lower case for our named results.
+	sql := `
+            SELECT COLUMN_NAME AS column_name FROM information_schema.STATISTICS AS index_cols INNER JOIN
+            (
+                SELECT stats.INDEX_NAME, SUM(
+                                              CASE LOWER(cols.DATA_TYPE)
+                                                WHEN 'enum' THEN 0
+                                                WHEN 'tinyint' THEN 1
+                                                WHEN 'year' THEN 2
+                                                WHEN 'smallint' THEN 3
+                                                WHEN 'date' THEN 4
+                                                WHEN 'mediumint' THEN 5
+                                                WHEN 'time' THEN 6
+                                                WHEN 'int' THEN 7
+                                                WHEN 'set' THEN 8
+                                                WHEN 'timestamp' THEN 9
+                                                WHEN 'bigint' THEN 10
+                                                WHEN 'float' THEN 11
+                                                WHEN 'double' THEN 12
+                                                WHEN 'decimal' THEN 13
+                                                WHEN 'datetime' THEN 14
+                                                WHEN 'binary' THEN 15
+                                                WHEN 'char' THEN 16
+                                                WHEN 'varbinary' THEN 17
+                                                WHEN 'varchar' THEN 18
+                                                WHEN 'tinyblob' THEN 19
+                                                WHEN 'tinytext' THEN 20
+                                                ELSE 1000
+                                              END
+                                            ) AS type_cost, COUNT(stats.COLUMN_NAME) AS col_count FROM information_schema.STATISTICS AS stats INNER JOIN
```

Comment:
Is the method of computing type cost something we/others use, or is this our heuristic? Using this logic, 2 int columns are better than one varchar column, but not 3 int columns, for example.

Reply:
It's our heuristic. It's an improved version of the one already used for OnlineDDL — where we only take the first column type into account and favor integer types over char types.

Reply:
I just tweaked it a bit. Thanks for pointing this case out as it got me thinking... 🙂

Reply:
I think the heuristic here is "good enough" and predicts a good result most of the time. This is just due to that fact that unique key columns are not randomly assorted by some fuzzer, they are created by human beings, and if the first column is an integer as opposed to a text, it's _likely_ that overall the index is gonna be better. I'm just speculating of course, I don't have numbers.
At any case, like @mattlord says, this is an improvement over the already existing query used in `vitess` migrations, which is derived from `gh-ost`, where it worked/works very well. In fact, I don't remember a `gh-ost` issue complaining about the chosen index. Let's tick with this heuristic!